sdk/core/System.BinaryData/src/BinaryData.cs

```diff
@@ -142,12 +142,11 @@ public BinaryData(string data)
                 streamLength = (int)longLength;
             }
 
-            MemoryStream? inputMemoryStream = stream as MemoryStream;
             ArraySegment<byte> buffer = default;
 
-            if (inputMemoryStream?.TryGetBuffer(out buffer) ?? false)
+            if (stream.GetType() == typeof(MemoryStream) && ((stream as MemoryStream)?.TryGetBuffer(out buffer) ?? false))
```

Comment:
needed to check exact type in case someone derives from MemoryStream and the assumptions we are making no longer hold, e.g. in the MaxStreamLength respected test case.